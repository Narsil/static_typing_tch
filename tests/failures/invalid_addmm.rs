use static_typing_tch::tensor_check;
use tch::Tensor;

tensor_check! {
fn gelu(x: &Tensor<(BS, H)>) -> Tensor<(BS, H)> {
    let y: Tensor<(BS, H)> = 0.79788456 * x * (1.0 + 0.044715 * x * x);
    x * 0.5 * (1.0 + y.tanh())
}

fn transformer_mlp(
    hidden_states: &Tensor<(BS, H)>,
    dense_h_to_4h: &Tensor<(H, H4)>,
    dense_h_to_4h_bias: &Tensor<(H4,)>,
    dense_4h_to_h: &Tensor<(H3, H)>, // Dimension mismatch it should be h3
    dense_4h_to_h_bias: &Tensor<(H,)>,
) -> Tensor<(B, S, H)> {
    let hidden_states = dense_h_to_4h_bias.addmm(&hidden_states, &dense_h_to_4h);
    let hidden_states = gelu(&hidden_states);
    let hidden_states = dense_4h_to_h_bias.addmm(&hidden_states, &dense_4h_to_h);
    hidden_states
}

fn transformer_mlp2(
    hidden_states: &Tensor<(BS, H)>,
    dense_h_to_4h: &Tensor<(H, H4)>,
    dense_h_to_4h_bias: &Tensor<(H4,)>,
    dense_4h_to_h: &Tensor<(H4, H)>,
    dense_4h_to_h_bias: &Tensor<(H,)>,
) -> Tensor<(B, S, H)> {
    let hidden_states = dense_h_to_4h_bias.addmm(&hidden_states, &dense_h_to_4h);
    let hidden_states = gelu(&hidden_states);
    let hidden_states = dense_4h_to_h_bias.addmm(&hidden_states, &dense_h_to_4h); // Dimension mismatch wrong argument

    hidden_states
}

fn transformer_mlp3(
    hidden_states: &Tensor<(B, S, H)>,
    dense_h_to_4h: &Tensor<(H, H4)>,
    dense_h_to_4h_bias: &Tensor<(H4,)>,
    dense_4h_to_h: &Tensor<(H4, H)>,
    dense_4h_to_h_bias: &Tensor<(H,)>,
) -> Tensor<(B, S, H)> {
    let hidden_states = dense_h_to_4h_bias.addmm(&hidden_states, &dense_h_to_4h); //Mismatch addmm
                                                                                  //supports only
                                                                                  //matrices, a
                                                                                  //`.view(-1, ..)`
                                                                                  //is missing
    let hidden_states = gelu(&hidden_states);
    let hidden_states = dense_4h_to_h_bias.addmm(&hidden_states, &dense_h_to_4h);

    hidden_states
}

fn main() {}
}
