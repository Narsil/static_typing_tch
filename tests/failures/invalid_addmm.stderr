error: Second dim was expected to be Symbol("H3") but found Symbol("H4")
  --> tests/failures/invalid_addmm.rs:19:50
   |
19 |     let hidden_states = dense_4h_to_h_bias.addmm(&hidden_states, &dense_4h_to_h);
   |                                                  ^^^^^^^^^^^^^^

error: Second dim was expected to be Symbol("H") but found Symbol("H4")
  --> tests/failures/invalid_addmm.rs:32:50
   |
32 |     let hidden_states = dense_4h_to_h_bias.addmm(&hidden_states, &dense_h_to_4h); // Dimension mismatch wrong argument
   |                                                  ^^^^^^^^^^^^^^

error: Second dim was expected to be Symbol("H") but found Symbol("H4")
  --> tests/failures/invalid_addmm.rs:32:66
   |
32 |     let hidden_states = dense_4h_to_h_bias.addmm(&hidden_states, &dense_h_to_4h); // Dimension mismatch wrong argument
   |                                                                  ^^^^^^^^^^^^^^

error: This tensor must be 2 dimensional but found TensorType { shape: [Symbol("B"), Symbol("S"), Symbol("H")], kind: Implicit, device: Implicit }
  --> tests/failures/invalid_addmm.rs:44:50
   |
44 |     let hidden_states = dense_h_to_4h_bias.addmm(&hidden_states, &dense_h_to_4h); //Mismatch addmm
   |                                                  ^^^^^^^^^^^^^^

error: Second dim was expected to be Symbol("H") but found Symbol("S")
  --> tests/failures/invalid_addmm.rs:44:50
   |
44 |     let hidden_states = dense_h_to_4h_bias.addmm(&hidden_states, &dense_h_to_4h); //Mismatch addmm
   |                                                  ^^^^^^^^^^^^^^

error: Second dim was expected to be Symbol("H") but found Symbol("H4")
  --> tests/failures/invalid_addmm.rs:50:50
   |
50 |     let hidden_states = dense_4h_to_h_bias.addmm(&hidden_states, &dense_h_to_4h);
   |                                                  ^^^^^^^^^^^^^^

error: Second dim was expected to be Symbol("H") but found Symbol("H4")
  --> tests/failures/invalid_addmm.rs:50:66
   |
50 |     let hidden_states = dense_4h_to_h_bias.addmm(&hidden_states, &dense_h_to_4h);
   |                                                                  ^^^^^^^^^^^^^^
